apps/examples/board_specific_demo/ameba_secure_storage_demo/secure_storage.c

```diff
@@ -63,88 +69,150 @@ int sstorage_main(int argc, char *argv[])
+	snprintf(ss_path, 7, "ss/%d", slot_index);
+
+	printf("Let us read once before writing to the slot\n");
+	printf("ss_read_secure_storage() Slot: %d\n\n", slot_index);
+	res = ss_read_secure_storage(hnd, ss_path, 0, &data);
+	if (res != SECURITY_OK) {
+		printf("ss_read_secure_storage() Fail, %d, res = %d\n", __LINE__, res);
+	} else {
+		printf("data.length: %d\n", data.length);
+		memcpy(pre_read, data.data, data.length);
+		printf("data.data:\n");
+		for (i = 0; i < data.length; i++) {
+			printf("%02x ", pre_read[i]);
+			if (!((i + 1) % 16)) {
+				printf("\n");
+			}
 		}
 	}
 
-	printf("Start Secure Storage read!!\n");
 	sleep(1);
-	for (slot = 0; slot < TEST_SLOT_NUM; slot++) {
-		snprintf(ss_path, 7, "ss/%d", slot);
-		printf("ss_read_secure_storage() Slot: %d\n", slot);
-		sleep(1);
-		res = ss_read_secure_storage(hnd, ss_path, 0, &data);
-		if (res != SECURITY_OK) {
-			printf("ss_read_secure_storage() Fail\n", __LINE__);
-		} else {
-			printf("data.length: %d\n", data.length);
-			printf("data.data:\n");
-			memcpy(b, data.data, data.length);
-			for (int i = 0; i < data.length; i++) {
-				printf("%02x ", b[i]);
-				if (!((i + 1) % 16)) {
-					printf("\n");
-				}
+	printf("\nStart Flash Raw data read!!\n");
+	/* The Secure Storage Slot Flash Address , 156 = HEADER 4 + TEST_DATA_LENGTH 4 + DATA LENGTH + TAG 16 + FOOTER 4 , 1 = en_display*/
+	length = HEAD_LENGTH + DATA_LENGTH + TEST_DATA_LENGTH + TAG_LENGTH + HEAD_LENGTH;
+	if (slot_index == 0) {
+		address = SLOT0_START_ADDR;
+	} else {
+		address = SLOT1_START_ADDR + (SLOT_SIZE * (slot_index - 1));
+	}
+
+	printf("Read Raw data from Slot: %d, Area A\n", slot_index);
+	if (!rtl_ss_flash_read(address, length, pre_read, 1)) {
+		printf("rtl_ss_flash_read() Fail, %d\n", __LINE__);
+	}
+
+	sleep(1);
+	address += SLOT_AREA_OFFEST;
+	printf("\nRead Raw data from Slot: %d, Area B\n", slot_index);
+	if (!rtl_ss_flash_read(address, length, pre_read, 1)) {
+		printf("rtl_ss_flash_read() Fail, %d\n", __LINE__);
+	}
+
+	printf("Read complete\n");
+	sleep(1);
+
+	printf("Start Secure Storage write!!\n");
+	/* Populate the buffer with the user-given data */
+	memset(a, input_data, TEST_DATA_LENGTH);
+	data.data = (void *)a;
+	printf("ss_write_secure_storage() Slot: %d\n", slot_index);
+
+	/* Write TEST_DATA_LENGTH bytes from buffer 'a' to the desired slot of the SS */
+	res = ss_write_secure_storage(hnd, ss_path, 0, &data);
+	if (res != SECURITY_OK) {
+		printf("ss_write_secure_storage() Fail, %d, res = %d\n", __LINE__, res);
+	}
+	printf("Write Complete\n");
+	sleep(1);
+
+	printf("\nStart Secure Storage read!!\n");
+	printf("ss_read_secure_storage() Slot: %d\n\n", slot_index);
+	res = ss_read_secure_storage(hnd, ss_path, 0, &data);
+	if (res != SECURITY_OK) {
+		printf("ss_read_secure_storage() Fail, %d, res = %d\n", __LINE__, res);
+	} else {
+		printf("data.length: %d\n", TEST_DATA_LENGTH);
+		memcpy(post_read, data.data, data.length);
```

Comment:
same as previous

Reply:
All these 3 comments are related to the same concern that I raised.
Ss_secure_storage_read function allocates buffer of its own but never frees it.

In the application, after we do memcpy from.data.data to our app buffer, we can free data.data to prevent memory leaks in this app. Is that what you suggest?

However, it is not standard practice for app dev to free memory that he/she never allocated. Thus it would be good if the API is made standard to accept a data buffer from the app and populate that.

Reply:
1. It looks no need to copy data.data into local buffer. We can print data.data directly.
2. The read function allocates a memory and it should be freed in the app. But after using, before calling read function again, it should be freed to prevent memory leak. But it doesn't. At the end of the app, there is a single free for data.data. We should free it before using read function again

Reply:
I agree, point 1 is better app design, uses less memory.

For point 2, yes this is what I also think can be the solution here. 
However, I just want to confirm that we are on the same page - asking app developer to free a buffer that they never allocated is not standard.
 But for now, since we have to work with Realtek's current design, we will do it this way?

Reply:
Even I agree what you said, it's not a TizenRT policy. That's why security guy did not design like that.
When we get new security engineer, let's discuss it then. If we want to modify it, it should be concern much large parts like app (to free newly), security hal layer for entire chipsets and so on.
It's not a part of this PR. This PR makes the example working without any issue.